man/socklog-check.8
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "socklog\-check" "8" 
.SH NAME
socklog\-check \- checks for the availability of a socklog(8) service.
.SH SYNOPSIS
\f[B]socklog\-check\f[R] [\-v] [unix [\f[I]address\f[R]]]
.SH DESCRIPTION
\f[B]socklog\-check\f[R] attempts to connect to a \f[B]socklog\f[R](8)
unix service and reports whether the connection attempt succeeded or
not.
.PP
It connects to the datagram oriented unix domain socket
\f[I]address\f[R], and reports success through the return code.
If \f[I]address\f[R] is not specified, \f[B]socklog\-check\f[R] attempts
to connect to \f[I]/dev/log\f[R].
.PP
When running services that use the syslog facility under runit\[aq]s
service supervision, \f[B]socklog\-check\f[R] should be added to the top
of the service\[aq]s run script to ensure that no syslog messages are
generated until a syslog service is up and running to process the
messages, e.g.:
.IP
.EX
#!/bin/sh
socklog\-check || exit 1
exec service_using_syslog
.EE
.SH OPTIONS
.TP
\f[B]\-v\f[R]
verbose.
Print verbose message to standard error.
.SH EXIT CODES
On success \f[B]socklog\-check\f[R] returns 0.
.PP
On error, \f[B]socklog\-check\f[R] prints a message to standard error
and exits 111.
.SH SEE ALSO
socklog(8), socklog\-conf(8), svlogd(8), tryto(1), uncat(1), sv(8),
runsv(8), runsvdir(8)
.PP
https://smarden.org/socklog/
.PD 0
.P
.PD
https://smarden.org/runit/
.SH AUTHOR
Gerrit Pape <pape\[at]smarden.org>


package/CHANGES
  * doc/index.html: spelling.
  * doc/install.html: how to install as non-root user.
  * doc/examples.html, doc/notify.html: upper and lower case, format.
  * doc/*.html, man/*.?: replace; converted from markdown source.
  * socklog-check.c, socklog.c: check length of unix socket to match struct
    sockaddr.
  * pathexec_env.c, socklog-conf.c, socklog.c, tryto.c, uncat.c: whitespace.
  * socklog.c: fix compilation warning -Wpointer-sign.
  * socklog.c, man/socklog.8: add -T to drop timestamps from syslog messages
    (github/leahneukirchen).
  * buffer_put.c, byte.h, byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c,
    byte_rchr.c, byte_zero.c, str_chr.c, str_len.c: sync daemontools library
    fork with runit source.

socklog 2.1.2
Sun, 02 Nov 2025 11:41:11 +0100
  * alloc.c, alloc_re.c, buffer_0.c, byte_chr.c, byte_copy.c, byte_cr.c,
    byte_diff.c, byte_rchr.c, byte_zero.c, error.h, select.h2, wait.h,
    wait_nohang.c, wait_pid.c: sync daemontools library fork with runit
    source.
  * doc/*.html: !DOCTYPE.
  * socklog.c, tryto.c, uncat.c: use parameter name "unused" in sighandler
    functions.
  * hasflock.h1: add #include <unistd.h>.
  * alloc.h, buffer.c, buffer.h, buffer_get.c, buffer_put.c, buffer_write.c,
    byte.h, check-socklog-unix.c, pathexec_env.c, select.h1, select.h2,
    sig.c, sig.h, sig_catch.c, socklog-check.c, socklog-check.dist,
    socklog.c, tryto.c, tryto.dist, uncat.c, uncat.dist, wait.h: fix build
    with gcc15: in C23 f() means f(void), consequently signal handler
    functions now are f(int), alloc_free() and byte_zero() parameters need
    casts, buffer_unixwrite() removes "const" from parameter.
  * package/README: update copyright notice.
  * doc/configuration.html: replace links to daemontools faq with runit faq.
  * doc/*: switch all cr.yp.to http links to https.
  * doc/*, man/*, package/README: switch all smarden.org http links to https.
  * tryto.c: don't report failure if the final try of prog succeeds (thx
    Andras Korn).
  * doc/index.html: remove reference to mailing list misc@list.smarden.org.
  * doc/index.html: add reference to repository on github.com.

socklog 2.1.1
Fri, 11 Oct 2024 00:00:17 +0000
  * doc/install.html: add link to sha256sum.asc.
  * doc/install.html: change mail address.
  * doc/index.html: remove reference to socklog 2.0.x.
  * doc/*.html: remove <small>$Id$</small>.
  * debian/: remove; obsolete.
  * pathexec.h, pathexec_env.c, pathexec_run.c, seek_set.c, sgetopt.c,
    sgetopt.h, socklog-check.c, socklog-conf.c, socklog.c, subgetopt.c,
    subgetopt.h, trycpp.c, tryflock.c, trypoll.c, trysgact.c, trysgprm.c,
    tryto.c, tryulong64.c, trywaitp.c, uncat.c, x86cpuid.c: incorporate
    lib changes from runit-2.2.0 and adapt socklog programs to properly
    build with modern toolchains (thx Gentoo, Debian, Void Linux).
  * prot.c, prot.h, chkshsgr.c, tryshsgr.c, hasshsgr.h?, warn-shsgr:
    remove; obsolete.
  * Makefile, TARGETS: no longer check for "shortsetgroups" (thx Leah
    Neukirchen).
  * socklog.c: replace prot_gid() with setgroups() and prot_uid() with
    setuid().

socklog 2.1.0
Mon, 06 Mar 2006 12:59:33 +0000
  * socklog-conf, man/socklog-conf.8: create service directories
    /etc/sv/socklog-* instead of /etc/socklog/*; create ./check script
    for socklog-unix service.
  * doc/configuration.html, doc/install.html, doc/network.html,
    doc/notify.html: adapt to new path to service directories; utilize
    runit's sv program instead of runsvctrl.
  * man/socklog-check.8, man/socklog-conf.8, man/socklog.8, man/uncat.1:
    see also sv(8).

socklog 2.0.3
Sun, 26 Feb 2006 13:29:17 +0000
  * doc/index.html: minor.
  * socklog-check.c: change EDESTADDRREQ to ECONNREFUSED if defined and set
    by connect().
  * tryto: fix a race with child terminating immediatly (thx Alex Efros);
    catch SIGCHLD while waiting ktimeout seconds before sending KILL.
  * man/tryto.1: remove reference to timelimit(8) (thx Alex Efros).
  * socklog.check, socklog.dist: don't check usage output with -V as it
    differs on solaris (thx Asif Iqbal).

socklog 2.0.2
Sun, 10 Apr 2005 07:21:30 +0000
  * doc/readme.solaris.html: add -ldoor -lthread to src/conf-ld (thx Lars
    Kellogg-Stedman); update log/run script, mailing list address.
  * man/socklog.8: remove double entries in SEE ALSO.

socklog 2.0.1
Thu, 10 Feb 2005 10:05:55 +0000
  * Makefile: add linker options detected by trysocketlib.c to targets
    check-socklog-inet, check-socklog-inet (fixes test failure on solaris,
    thx Uffe Jakobsen).

socklog 2.0.0
Sun, 06 Feb 2005 19:57:04 +0000
  * socklog-check.check: tweak timing in program test.
  * tryto.c: new option -P: run prog in new session, and signal process
    group.
  * man/tryto.1: adapt; cleanup.
  * man/socklog-check.8: add hint about runit services using syslog
    facility; cleanup.
  * man/socklog-conf.8, man/socklog.8, man/uncat.1: cleanup.
  * doc/index.html: new stable release.
  * doc/network.html: typo.

socklog 1.5.0
Sat, 26 Jun 2004 11:28:33 +0000
  * socklog-check, man/socklog-check.8: new; check for the availability of
    a socklog service.
  * socklog-check.check, socklog-check.dist: new; program test.
  * byte_zero.c: new.
  * socklog.c, man/socklog.8: new option -U: don't override umask when
    creating domain socket.
  * package/check, package/compile, package/install-man, package/upgrade:
    minor cleanup.
  * Makefile, man/socklog-conf.8: adapt.

socklog 1.4.2
Tue, 23 Mar 2004 18:16:17 +0000
  * doc/index.html, doc/install.html: minor.
  * socklog.c: fix sigterm handling in ucspi mode.
  * uncat.check: tweak timing in program test.
  * Makefile: don't run target check by default.
  * package/check: new; run checks on programs.
  * package/install: run package/check.

socklog 1.4.1
Sat, 28 Feb 2004 17:47:13 +0000
  * socklog.c: handle sigterm on other platforms as accurately as already
    done on solaris.
  * tryto.c, uncat.c: minor cleanup.

src/socklog-check.c
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>
#include "strerr.h"
#include "str.h"
#include "byte.h"
#include "sgetopt.h"

#define FATAL "socklog-check: fatal: "
#define WARNING "socklog-check: warning: "
#define INFO "socklog-check: info: "

#define VERSION "$Id$"
#define USAGE " [-v] [unix [address]]"

const char *progname;

void usage() { strerr_die4x(1, "usage: ", progname, USAGE, "\n"); }

int s;
struct sockaddr_un sa;
const char *address ="/dev/log";
unsigned int verbose =0;

int main(int argc, char **argv) {
  int opt;

  progname =*argv;

  while ((opt =getopt(argc, argv, "vV")) != opteof) {
    switch(opt) {
    case 'v': verbose =1; break;
    case 'V': strerr_warn1(VERSION, 0);
    case '?': usage();
    }
  }
  argv +=optind;

  if (argv && *argv) {
    if (**argv != 'u') usage();
    if (++argv && *argv) address =*argv;
  }
  if ((s =str_len(address)) >= sizeof(sa.sun_path))
    strerr_die4x(111, FATAL, "unable to create socket: ", address,
                 ": name too long");
  byte_zero((char *)&sa, sizeof(sa));
  sa.sun_family =AF_UNIX;
  byte_copy(sa.sun_path, s, address);
  if ((s =socket(AF_UNIX, SOCK_DGRAM, 0)) == -1)
    strerr_die4sys(111, FATAL, "unable to create socket: ", address, ": ");
  if (connect(s, (struct sockaddr *)&sa, sizeof(sa)) == -1) {
    close(s);
#ifdef EDESTADDRREQ
    if (errno == EDESTADDRREQ) errno =error_connrefused;
#endif
    strerr_die4sys(111, WARNING, "unable to connect socket: ", address, ": ");
  }
  close(s);
  if (verbose) strerr_die3x(0, INFO, "successfully connected to ", address);
  return(0);
}
